File: src/BFS/WordLadder.java
package BFS;

import DFS.NumberOfIslands;

import java.lang.reflect.Array;
import java.util.*;

// https://leetcode.com/problems/word-ladder/
public class WordLadder {
    class Pair {
        String word;
        int step;

        public Pair(String word, int step) {
            this.word = word;
            this.step = step;
        }
    }

    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        int steps = 0;
        if (!wordList.contains(endWord)) {
            return steps;
        }

        Queue<Pair> bfsQueue = new LinkedList<>();
        boolean[] visited = new boolean[wordList.size()];

        bfsQueue.add(new Pair(beginWord, 1));

        while (!bfsQueue.isEmpty()) {
            Pair wordStep = bfsQueue.poll();

            if (endWord.equals(wordStep.word)) {
                System.out.print(wordStep.word + "\n");
                steps = wordStep.step;
                break;
            }
            System.out.print(wordStep.word+" -> ");

            for (int i = 0; i < wordList.size(); i++) {
                if (!visited[i] && isNeighbor(wordStep.word, wordList.get(i))) {
                    bfsQueue.add(new Pair(wordList.get(i), wordStep.step + 1));
                    visited[i] = true;
                }
            }
        }

        return steps;
    }

    public int ladderLength1(String beginWord, String endWord, List<String> wordList) { // level
        int steps = 0;
        if (!wordList.contains(endWord)) {
            return steps;
        }

        Queue<String> bfsQueue = new LinkedList<>();
        boolean[] visited = new boolean[wordList.size()];

        bfsQueue.add(beginWord);

        while (!bfsQueue.isEmpty()) {
            int count = bfsQueue.size();
            for (int j = 0; j < count; j++) {
                String word = bfsQueue.poll();
                if (endWord.equals(word)) {
                    System.out.print(word + "\n");
                    return steps  + 1;
                }
                System.out.print(word+" -> ");

                for (int i = 0; i < wordList.size(); i++) {
                    if (!visited[i] && isNeighbor(word, wordList.get(i))) {
                        bfsQueue.add(wordList.get(i));
                        visited[i] = true;
                    }
                }
            }
            steps++;
        }

        return 0;
    }

    boolean isNeighbor(String word, String anotherWord) {
        int numberOfCommonLetters = 0;
        for (int i = 0; i < word.length(); i++) {
            if (word.charAt(i) == anotherWord.charAt(i)) {
                numberOfCommonLetters++;
            }
        }
        return word.length() - numberOfCommonLetters == 1;
    }

    // Driver method
    public static void main(String[] args) throws java.lang.Exception {
        ArrayList<String> wordList = new ArrayList<>(Arrays.asList("hot","dot","dog","lot","log","cog"));
        String beginWord = "hit";
        String endWord = "cog";

        WordLadder w = new WordLadder();
        System.out.println("Number of steps is: " + w.ladderLength(beginWord, endWord, wordList));
        System.out.println("Number of steps is: " + w.ladderLength1(beginWord, endWord, wordList));
    }
}

File: py/essential/WordLadder.py
# https://leetcode.com/problems/word-ladder/description/?envType=problem-list-v2&envId=rab78cw1
# Grind
# Neet

from collections import deque
from itertools import product
from typing import List


class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        def isAdjacent(word1, word2):
            count = 0
            for i in range(len(word1)):
                if word1[i] != word2[i]:
                    count += 1
            return count == 1

        n = len(beginWord)
        res = 0
        visited = set() # to avoid a circle
        q = deque([beginWord])
        while q:
            qSize = len(q)
            for _ in range(qSize):
                curWord = q.popleft()
                if curWord == endWord:
                    return res + 1
                for word in wordList:
                    if word not in visited and isAdjacent(curWord, word):
                        visited.add(word)
                        q.append(word)
            res += 1
        return 0


s = Solution()
beginWord = "hit"
endWord = "cog"
wordList = ["hot", "dot", "dog", "lot", "log", "cog"]
print(s.ladderLength(beginWord, endWord, wordList))
beginWord = "hit"
endWord = "cog"
wordList = ["hot", "dot", "dog", "lot", "log"]
print(s.ladderLength(beginWord, endWord, wordList))
